lib/eslint.js

```diff
@@ -958,7 +958,7 @@ module.exports = (function() {
             // if current node is function declaration, add it to the list
             var current = controller.current();
             if (["FunctionDeclaration", "FunctionExpression",
-                    "ArrowFunctionExpression"].indexOf(current.type) >= 0) {
+                    "ArrowFunctionExpression", "SwitchStatement"].indexOf(current.type) >= 0) {
```

Comment:
Loops with fine because they have a BlockStatement that creates the scope. SwitchStatement has no block statement in the AST, so it needs an exception.